File: queue/queue.py
class Node:
  def __init__(self, data):
    self.val = data
    self.next = None

class Queue:
  def __init__(self):
    self.left = self.right = None

  # O(1) time complexity
  def enqueue(self, data):
    new_node = Node(data)

    if self.left:
      self.right.next = new_node
      self.right = new_node
    else:
      self.left = self.right = new_node

  def dequeue(self):
    if not self.left:
      raise ValueError("Empty queue")

    data = self.left.val
    self.left = self.left.next
    return data

  def printQueue(self):
    curr = self.left
    while curr:
      print(curr.val, "->")
      curr = curr.next

q = Queue()
q.enqueue(1)
q.enqueue(2)
q.enqueue(3)
q.enqueue(4)
q.dequeue()
q.printQueue()

File: queue/queue.cpp
#include <iostream>

class Node
{
public:
	Node(int value)
		: m_Value(value) {}
public:
	int m_Value;
	Node* m_Next = nullptr;
};

class Queue
{
public:
	void Enqueue(int value)
	{
		Node* newNode = new Node(value);

		if (!left)
		{
			left = newNode;
			right = newNode;
			return;
		}

		newNode->m_Next = left;
		left = newNode;
	}
	void Dequeue()
	{
		if (!left) return;
		Node* tempNode = left;
		left = left->m_Next;
		delete tempNode;
	}
	void Print()
	{
		Node* current = left;
		while (current)
		{
			std::cout << current->m_Value << "\n";
			current = current->m_Next;
		}
	}
private:
	Node* left = nullptr;
	Node* right = nullptr;
};

int main()
{
	Queue queue;
	queue.Enqueue(1);
	queue.Enqueue(2);
	queue.Enqueue(3);
	queue.Dequeue();
	queue.Print();
}